servicetalk-concurrent-api/src/testFixtures/java/io/servicetalk/concurrent/api/TestCollectingPublisherSubscriber.java

```diff
@@ -0,0 +1,213 @@
+import io.servicetalk.concurrent.PublisherSource.Subscription;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.internal.ThrowableUtils.unknownStackTrace;
+
+/**
+ * A {@link Subscriber} that enqueues {@link #onNext(Object)} and terminal signals while providing blocking methods
+ * to consume these events. There are two approaches to using this class:
+ * <pre>
+ *     TestCollectingPublisherSubscriber&lt;String&gt; sub = new TestCollectingPublisherSubscriber&lt;&gt;();
+ *
+ *     // Approach 1 - verify individual items sequentially.
+ *     String s = sub.takeOnNext();
+ *     // verify s
+ *     sub.awaitOnComplete(); // this will verify that all onNext signals have been consumed
+ *
+ *     // Approach 2 - wait for terminal, verify items in bulk.
+ *     sub.awaitOnComplete(false); // wait for the terminal signal, ignore if there are unconsumed onNext signals.
+ *     List&lt;String&gt; onNextSignals = sub.pollAllOnNext();
+ *     // verify all onNextSignals occurred in the expected order
+ * </pre>
+ * @param <T> The type of data in {@link #onNext(Object)}.
+ */
+public final class TestCollectingPublisherSubscriber<T> implements Subscriber<T> {
+    private static final Object NULL_ON_NEXT = new Object();
+    private static final Throwable ON_COMPLETE = unknownStackTrace(new RuntimeException("onComplete"),
+            TestCollectingPublisherSubscriber.class, "onComplete");
+    private final BlockingQueue<Object> items = new LinkedBlockingQueue<>();
+    private final CountDownLatch onTerminalLatch = new CountDownLatch(1);
+    private final CountDownLatch onSubscribeLatch = new CountDownLatch(1);
+    @Nullable
+    private Throwable onTerminal;
+    @Nullable
+    private Subscription subscription;
+
+    @Override
+    public void onSubscribe(final Subscription subscription) {
+        this.subscription = subscription;
+        onSubscribeLatch.countDown();
+    }
+
+    @Override
+    public void onNext(@Nullable final T t) {
+        items.add(t == null ? NULL_ON_NEXT : t);
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+        onTerminal = t;
+        onTerminalLatch.countDown();
+    }
+
+    @Override
+    public void onComplete() {
+        onTerminal = ON_COMPLETE;
+        onTerminalLatch.countDown();
+    }
+
+    /**
+     * Block until {@link #onSubscribe(Subscription)}.
+     *
+     * @return The {@link Subscription} from {@link #onSubscribe(Subscription)}.
+     * @throws InterruptedException if an interrupt occurs while blocking for waiting for
+     * {@link #onSubscribe(Subscription)}.
+     */
+    public Subscription awaitSubscription() throws InterruptedException {
+        onSubscribeLatch.await();
+        if (subscription == null) {
```

Comment:
Should we store `subscription` in a local var so that we are null-checking and returning the same value?

Alternatively, add an `assert subscription == null;` in `onSubscribe()`.

Reply:
I shifted the null check to the `onSubscribe`, made sure the value wasn't set twice, and changed this to an assert. PTAL